include/onnxruntime/core/platform/threadpool.h

```diff
@@ -1,95 +1,267 @@
+  double compute_cycles_;
+};
 
-/**
- * Generic class for instantiating thread pools.
- * Don't put any object of this type into a global variable in a Win32 DLL.
- */
+template <typename Environment>
+class ThreadPoolTempl;
+namespace concurrency {
 class ThreadPool {
  public:
-  /*
-  Initializes a thread pool given the current environment.
-  */
-  ThreadPool(const std::string& name, int num_threads);
-
-  /*
-  Enqueue a unit of work.
-  */
+  // Scheduling strategies for ParallelFor. The strategy governs how the given
+  // units of work are distributed among the available threads in the
+  // threadpool.
+  enum class SchedulingStrategy {
+    // The Adaptive scheduling strategy adaptively chooses the shard sizes based
+    // on the cost of each unit of work, and the cost model of the underlying
+    // threadpool device.
+    //
+    // The 'cost_per_unit' is an estimate of the number of CPU cycles (or
+    // nanoseconds if not CPU-bound) to complete a unit of work. Overestimating
+    // creates too many shards and CPU time will be dominated by per-shard
+    // overhead, such as Context creation. Underestimating may not fully make
+    // use of the specified parallelism, and may also cause inefficiencies due
+    // to load balancing issues and stragglers.
+    kAdaptive,
+    // The Fixed Block Size scheduling strategy shards the given units of work
+    // into shards of fixed size. In case the total number of units is not
+    // evenly divisible by 'block_size', at most one of the shards may be of
+    // smaller size. The exact number of shards may be found by a call to
+    // NumShardsUsedByFixedBlockSizeScheduling.
+    //
+    // Each shard may be executed on a different thread in parallel, depending
+    // on the number of threads available in the pool. Note that when there
+    // aren't enough threads in the pool to achieve full parallelism, function
+    // calls will be automatically queued.
+    kFixedBlockSize
+  };
+
+  // Contains additional parameters for either the Adaptive or the Fixed Block
+  // Size scheduling strategy.
+  class SchedulingParams {
+   public:
+    explicit SchedulingParams(SchedulingStrategy strategy, optional<int64_t> cost_per_unit,
+                              optional<int64_t> block_size)
+        : strategy_(strategy), cost_per_unit_(cost_per_unit), block_size_(block_size) {
+    }
+
+    SchedulingStrategy strategy() const {
+      return strategy_;
+    }
+    optional<int64_t> cost_per_unit() const {
+      return cost_per_unit_;
+    }
+    optional<int64_t> block_size() const {
+      return block_size_;
+    }
+
+   private:
+    // The underlying Scheduling Strategy for which this instance contains
+    // additional parameters.
+    SchedulingStrategy strategy_;
+
+    // The estimated cost per unit of work in number of CPU cycles (or
+    // nanoseconds if not CPU-bound). Only applicable for Adaptive scheduling
+    // strategy.
+    optional<int64_t> cost_per_unit_;
```

Comment:
>int64_t [](start = 13, length = 7)

We could always designate negative value as its absense

Reply:
Then no need to use optional?

Reply:
The code is copied from Tensorflow, I'm a little bit to prefer to keep it as it is. @pranavsharma What do you think?